File: Objective-C/TCMPPDemo/TCMPPSubscribeInfoVC.m
//
//  TCMPPSubscribeInfoVC.m
//  TCMPPDemo
//
//  Created by Assistant on 2024/12/19.
//  Copyright © 2024 Tencent. All rights reserved.
//

#import "TCMPPSubscribeInfoVC.h"
#import "TCMPPDemoLoginManager.h"
#import <TCMPPSDK/TCMPPSDK.h>
#import "ToastView.h"

#define kAppIconViewWidth 25.0f
#define kAppIconViewHeight 25.0
#define kAppNameLblLeftMargin 10.0f
#define kTitleLblTopMargin 10.0f
#define kContentLblTopMargin 5.0f
#define kCheckDetailsLblHeight 20.0f
#define kSeparatorLineHeight 0.5f
#define kContentLblFontSize 14.0f

@interface SubscribeInfoModel : NSObject

@property (nonatomic, copy) NSString *TmplID;
@property (nonatomic, copy) NSString *Content;
@property (nonatomic, assign) NSInteger DataTime;
@property (nonatomic, copy) NSString *MessageID;
@property (nonatomic, copy) NSString *TmplTitle;
@property (nonatomic, copy) NSString *MnpName;
@property (nonatomic, copy) NSString *MnpId;
@property (nonatomic, copy) NSString *Page;
@property (nonatomic, copy) NSString *State;
@property (nonatomic, assign) CGFloat cellH;

- (SubscribeInfoModel *)initWithJsonDatas:(NSDictionary *)jsonDatas;

@end

@implementation SubscribeInfoModel

- (SubscribeInfoModel *)initWithJsonDatas:(NSDictionary *)jsonDatas {
    if (self = [super init]) {
        self.TmplID = jsonDatas[@"TmplID"];
        self.Content = jsonDatas[@"Content"];
        self.DataTime = [jsonDatas[@"DataTime"] integerValue];
        self.MessageID = jsonDatas[@"MessageID"];
        self.TmplTitle = jsonDatas[@"TmplTitle"];
        self.MnpName = jsonDatas[@"MnpName"];
        self.MnpId = jsonDatas[@"MnpId"];
        self.Page = jsonDatas[@"Page"];
        self.State = jsonDatas[@"State"];
    }
    return self;
}

@end

@interface SubscribeInfoCell : UITableViewCell

@property (nonatomic, strong) UILabel *timeLbl;
@property (nonatomic, strong) UIView *cardView;
@property (nonatomic, strong) UIImageView *appIconView;
@property (nonatomic, strong) UILabel *appNameLbl;
@property (nonatomic, strong) UILabel *titleLbl;
@property (nonatomic, strong) UILabel *contentLbl;
@property (nonatomic, strong) UILabel *checkDetailsLbl;
@property (nonatomic, strong) UIView *separatorLine;
@property (nonatomic, strong) UIView *separatorLine1;
@property (nonatomic, assign) CGFloat contentLblH;
@property (nonatomic, strong) UIImageView *topRightImg;
@property (nonatomic, strong) UIImageView *bottomRightImg;

- (void)setCellDatas:(SubscribeInfoModel *)subscribeInfoModel;

@end

@implementation SubscribeInfoCell

- (instancetype)initWithStyle:(UITableViewCellStyle)style reuseIdentifier:(NSString *)reuseIdentifier {
    self = [super initWithStyle:style reuseIdentifier:reuseIdentifier];
    if (self) {
        [self setupSubviews];
    }
    return self;
}

- (void)setupSubviews {
    self.timeLbl = [[UILabel alloc] init];
    self.timeLbl.font = [UIFont boldSystemFontOfSize:14];
    self.timeLbl.textColor = [UIColor lightGrayColor];
    self.timeLbl.textAlignment = NSTextAlignmentCenter;
    [self.contentView addSubview:self.timeLbl];
    
    self.contentView.backgroundColor = [UIColor colorWithRed:234/255.0 green:234/255.0 blue:234/255.0 alpha:1];
    self.cardView = [[UIView alloc] init];
    self.cardView.backgroundColor = [UIColor whiteColor];
    self.cardView.layer.cornerRadius = 5.0;
    self.cardView.layer.masksToBounds = YES;
    [self.contentView addSubview:self.cardView];
    
    self.appIconView = [[UIImageView alloc] init];
    self.appIconView.image = [UIImage imageNamed:@"tcmpp_share_icon"];
    [self.cardView addSubview:self.appIconView];
    
    self.topRightImg = [[UIImageView alloc] init];
    self.topRightImg.image = [UIImage imageNamed:@"white_more"];
    [self.cardView addSubview:self.topRightImg];

    self.appNameLbl = [[UILabel alloc] init];
    self.appNameLbl.font = [UIFont boldSystemFontOfSize:16];
    self.appNameLbl.textColor = [UIColor blackColor];
    [self.cardView addSubview:self.appNameLbl];

    self.separatorLine = [[UIView alloc] init];
    self.separatorLine.backgroundColor = [UIColor lightGrayColor];
    [self.cardView addSubview:self.separatorLine];

    self.titleLbl = [[UILabel alloc] init];
    [self.cardView addSubview:self.titleLbl];

    self.contentLbl = [[UILabel alloc] init];
    self.contentLbl.numberOfLines = 0;
    self.contentLbl.textColor = [UIColor blackColor];
    self.contentLbl.font = [UIFont systemFontOfSize:kContentLblFontSize];
    [self.cardView addSubview:self.contentLbl];
    
    self.separatorLine1 = [[UIView alloc] init];
    self.separatorLine1.backgroundColor = [UIColor lightGrayColor];
    [self.cardView addSubview:self.separatorLine1];

    self.checkDetailsLbl = [[UILabel alloc] init];
    self.checkDetailsLbl.font = [UIFont systemFontOfSize:14];
    self.checkDetailsLbl.text = NSLocalizedString(@"Check the details", nil);
    self.checkDetailsLbl.textColor = [UIColor blackColor];
    [self.cardView addSubview:self.checkDetailsLbl];
    
    self.bottomRightImg = [[UIImageView alloc] init];
    self.bottomRightImg.image = [UIImage imageNamed:@"miniApp_tousu_icon"];
    [self.cardView addSubview:self.bottomRightImg];
}

- (void)layoutSubviews {
    [super layoutSubviews];
    
    self.timeLbl.frame = CGRectMake(0, 20, CGRectGetWidth(self.frame) - 20, 20);
    self.cardView.frame = CGRectMake(10, 50, CGRectGetWidth(self.frame) - 20, CGRectGetHeight(self.frame) - 50);
    self.appIconView.frame = CGRectMake(10, 10, kAppIconViewWidth, kAppIconViewHeight);
    self.appNameLbl.frame = CGRectMake(CGRectGetMaxX(self.appIconView.frame) + kAppNameLblLeftMargin, 13, CGRectGetWidth(self.frame) - CGRectGetMaxX(self.appIconView.frame) - kAppNameLblLeftMargin * 2, 20);
    self.topRightImg.frame = CGRectMake(self.frame.size.width - 70, 15, 35, 8);

    self.separatorLine.frame = CGRectMake(0, CGRectGetMaxY(self.appIconView.frame) + 10, CGRectGetWidth(self.cardView.frame), kSeparatorLineHeight);
    self.titleLbl.frame = CGRectMake(10, CGRectGetMaxY(self.separatorLine.frame) + kTitleLblTopMargin, CGRectGetWidth(self.frame) - 20, 20);
    self.contentLbl.frame = CGRectMake(10, CGRectGetMaxY(self.titleLbl.frame) + kContentLblTopMargin, CGRectGetWidth(self.frame) - 20, self.contentLblH);
    self.separatorLine1.frame = CGRectMake(10, CGRectGetMaxY(self.contentLbl.frame) + 10, CGRectGetWidth(self.cardView.frame) - 20, kSeparatorLineHeight);
    self.checkDetailsLbl.frame = CGRectMake(10, CGRectGetMaxY(self.separatorLine1.frame) + 10, 200, kCheckDetailsLblHeight);
    self.bottomRightImg.frame = CGRectMake(self.frame.size.width - 50, CGRectGetMaxY(self.separatorLine1.frame) + 10, 10, 15);
}

- (void)setCellDatas:(SubscribeInfoModel *)subscribeInfoModel {
    self.appNameLbl.text = subscribeInfoModel.MnpName;
    self.titleLbl.text = subscribeInfoModel.TmplTitle;
    self.contentLbl.text = subscribeInfoModel.Content;
    self.contentLblH = [self calculateContentLblHeight:subscribeInfoModel.Content];
    
    NSDate *date = [NSDate dateWithTimeIntervalSince1970:subscribeInfoModel.DataTime];
    NSDateFormatter *formatter = [[NSDateFormatter alloc] init];
    [formatter setDateFormat:@"yyyy-MM-dd HH:mm:ss"];
    self.timeLbl.text = [formatter stringFromDate:date];
}

- (CGFloat)calculateContentLblHeight:(NSString *)content {
    CGSize maxSize = CGSizeMake(CGRectGetWidth(self.frame) - 40, MAXFLOAT);
    CGSize contentSize = [content boundingRectWithSize:maxSize
                                              options:NSStringDrawingUsesLineFragmentOrigin
                                           attributes:@{NSFontAttributeName: [UIFont systemFontOfSize:kContentLblFontSize]}
                                              context:nil].size;
    return contentSize.height;
}

@end

@interface TCMPPSubscribeInfoVC () <UITableViewDataSource, UITableViewDelegate>

@property (nonatomic, strong) UITableView *tableView;
@property (nonatomic, strong) NSArray <SubscribeInfoModel *> *modelDatas;
@property (nonatomic, strong) UIRefreshControl *refreshControl;

@end

@implementation TCMPPSubscribeInfoVC

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.title = NSLocalizedString(@"Service Notice", nil);
    if (@available(iOS 13.0, *)) {
        self.view.backgroundColor = [UIColor systemBackgroundColor];
    } else {
        self.view.backgroundColor = [UIColor whiteColor];
    }
    
    [self setupTableView];
    [self requestDatas];
}

- (void)setupTableView {
    self.tableView = [[UITableView alloc] initWithFrame:self.view.bounds style:UITableViewStylePlain];
    self.tableView.dataSource = self;
    self.tableView.delegate = self;
    self.tableView.separatorStyle = UITableViewCellSeparatorStyleNone;
    self.tableView.backgroundColor = [UIColor colorWithRed:234/255.0 green:234/255.0 blue:234/255.0 alpha:1];
    [self.tableView registerClass:[SubscribeInfoCell class] forCellReuseIdentifier:@"SubscribeInfoCellID"];
    [self.view addSubview:self.tableView];
    
    self.tableView.translatesAutoresizingMaskIntoConstraints = NO;
    [NSLayoutConstraint activateConstraints:@[
        [self.tableView.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
        [self.tableView.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor],
        [self.tableView.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [self.tableView.bottomAnchor constraintEqualToAnchor:self.view.bottomAnchor]
    ]];
    
    self.refreshControl = [[UIRefreshControl alloc] init];
    [self.refreshControl addTarget:self action:@selector(refreshData) forControlEvents:UIControlEventValueChanged];
    self.tableView.refreshControl = self.refreshControl;
}

- (void)refreshData {
    [self requestDatas];
}

- (void)requestDatas {
    NSString *token = [TCMPPUserInfo sharedInstance].token;
    NSString *appId = [[TMFMiniAppSDKManager sharedInstance] getConfigAppKey];
    
    if (!token || !appId) {
        [self showToast:NSLocalizedString(@"Please login first", nil)];
        [self.refreshControl endRefreshing];
        return;
    }
    
    [[TCMPPDemoLoginManager sharedInstance] getMessage:token appId:appId offset:0 success:^(NSArray * _Nonnull message) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.refreshControl endRefreshing];
            
            if (![message isKindOfClass:[NSArray class]]) {
                [self showToast:NSLocalizedString(@"No messages", nil)];
                return;
            }
            
            if (!message.count) {
                [self showToast:NSLocalizedString(@"No messages", nil)];
                return;
            }
            
            NSMutableArray *arrayDatas = [NSMutableArray array];
            for (NSDictionary *json in message) {
                SubscribeInfoModel *model = [[SubscribeInfoModel alloc] initWithJsonDatas:json];
                [arrayDatas addObject:model];
            }
            self.modelDatas = [NSArray arrayWithArray:arrayDatas];
            [self.tableView reloadData];
        });
    } failure:^(NSError * _Nonnull error) {
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.refreshControl endRefreshing];
            [self showToast:error.localizedDescription ?: NSLocalizedString(@"Failed to load messages", nil)];
        });
    }];
}

#pragma mark - UITableViewDataSource

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.modelDatas.count;
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    SubscribeInfoCell *cell = [tableView dequeueReusableCellWithIdentifier:@"SubscribeInfoCellID" forIndexPath:indexPath];
    [cell setCellDatas:self.modelDatas[indexPath.row]];
    return cell;
}

#pragma mark - UITableViewDelegate

- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath {
    NSString *content = self.modelDatas[indexPath.row].Content;
    CGSize contentSize = [content boundingRectWithSize:CGSizeMake(self.view.frame.size.width - 40, CGFLOAT_MAX)
                                               options:NSStringDrawingUsesLineFragmentOrigin
                                            attributes:@{NSFontAttributeName: [UIFont systemFontOfSize:14]}
                                               context:nil].size;
    
    return contentSize.height + 180;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath {
    [tableView deselectRowAtIndexPath:indexPath animated:YES];
    
    SubscribeInfoModel *model = self.modelDatas[indexPath.row];
    TMAVersionType vertype = TMAVersionOnline;
    if ([model.State isEqualToString:@"developer"]) {
        vertype = TMAVersionDevelop;
    } else if ([model.State isEqualToString:@"trial"]) {
        vertype = TMAVersionPreview;
    }
    
    [[TMFMiniAppSDKManager sharedInstance] startUpMiniAppWithAppID:model.MnpId verType:vertype scene:0 firstPage:model.Page paramsStr:nil parentVC:self completion:^(NSError * _Nullable error) {
        if (error) {
            [self showErrorInfo:error];
            return;
        }
    }];
}

#pragma mark - Helper Methods
- (void)showErrorInfo:(NSError *)err {
    if (err == nil) {
        return;
    }
    NSString *errMsg = nil;
    NSString *localizedDescription = err.userInfo[NSLocalizedDescriptionKey];
    if (localizedDescription != nil) {
        errMsg = [NSString stringWithFormat:@"%@\n%ld\n%@", err.domain, (long)err.code, localizedDescription];
    } else {
        errMsg = [NSString stringWithFormat:@"%@\n%ld", err.domain, (long)err.code];
    }
    
    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Error" message:errMsg preferredStyle:UIAlertControllerStyleAlert];
    [self presentViewController:alert animated:YES completion:nil];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(2.0 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        [alert dismissViewControllerAnimated:YES completion:nil];
    });
}

- (void)showToast:(NSString *)message {
    UIImage *icon = [UIImage imageNamed:@"success"];
    ToastView *toast = [[ToastView alloc] initWithIcon:icon title:message];
    [toast showWithDuration:2.0];
}

@end 

File: Swift/TCMPPDemo-Swift/Class/TCMPPSubscribeInfoVC.swift
//
//  TCMPPSubscribeInfoVC.swift
//  TCMPPDemo-Swift
//
//  Created by Assistant on 2024/12/19.
//  Copyright © 2024 Tencent. All rights reserved.
//

import UIKit
import TCMPPSDK

struct SubscribeInfoModel {
    let tmplID: String
    let content: String
    let dataTime: Int
    let messageID: String
    let tmplTitle: String
    let mnpName: String
    let mnpId: String
    let page: String
    let state: String
    
    init(jsonData: [String: Any]) {
        self.tmplID = jsonData["TmplID"] as? String ?? ""
        self.content = jsonData["Content"] as? String ?? ""
        self.dataTime = jsonData["DataTime"] as? Int ?? 0
        self.messageID = jsonData["MessageID"] as? String ?? ""
        self.tmplTitle = jsonData["TmplTitle"] as? String ?? ""
        self.mnpName = jsonData["MnpName"] as? String ?? ""
        self.mnpId = jsonData["MnpId"] as? String ?? ""
        self.page = jsonData["Page"] as? String ?? ""
        self.state = jsonData["State"] as? String ?? ""
    }
}

class SubscribeInfoCell: UITableViewCell {
    private let timeLabel = UILabel()
    private let cardView = UIView()
    private let appIconView = UIImageView()
    private let appNameLabel = UILabel()
    private let titleLabel = UILabel()
    private let contentLabel = UILabel()
    private let checkDetailsLabel = UILabel()
    private let separatorLine = UIView()
    private let separatorLine1 = UIView()
    private let topRightImg = UIImageView()
    private let bottomRightImg = UIImageView()
    
    override init(style: UITableViewCell.CellStyle, reuseIdentifier: String?) {
        super.init(style: style, reuseIdentifier: reuseIdentifier)
        setupSubviews()
    }
    
    required init?(coder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }
    
    private func setupSubviews() {
        backgroundColor = UIColor(red: 234/255.0, green: 234/255.0, blue: 234/255.0, alpha: 1)
        selectionStyle = .none
        
        // Time label
        timeLabel.font = UIFont.boldSystemFont(ofSize: 14)
        timeLabel.textColor = UIColor.lightGray
        timeLabel.textAlignment = .center
        contentView.addSubview(timeLabel)
        
        // Card view
        cardView.backgroundColor = UIColor.white
        cardView.layer.cornerRadius = 5.0
        cardView.layer.masksToBounds = true
        contentView.addSubview(cardView)
        
        // App icon
        if #available(iOS 13.0, *) {
            appIconView.image = UIImage(systemName: "app.fill")
        } else {
            appIconView.image = UIImage(named: "tmf_weapp_icon_default")
        }
        cardView.addSubview(appIconView)
        
        // Top right image
        if #available(iOS 13.0, *) {
            topRightImg.image = UIImage(systemName: "ellipsis")
        } else {
            topRightImg.image = UIImage(named: "more")
        }
        cardView.addSubview(topRightImg)
        
        // App name
        appNameLabel.font = UIFont.boldSystemFont(ofSize: 16)
        appNameLabel.textColor = UIColor.black
        cardView.addSubview(appNameLabel)
        
        // Separator line
        separatorLine.backgroundColor = UIColor.lightGray
        cardView.addSubview(separatorLine)
        
        // Title
        cardView.addSubview(titleLabel)
        
        // Content
        contentLabel.numberOfLines = 0
        contentLabel.textColor = UIColor.black
        contentLabel.font = UIFont.systemFont(ofSize: 14)
        cardView.addSubview(contentLabel)
        
        // Separator line 1
        separatorLine1.backgroundColor = UIColor.lightGray
        cardView.addSubview(separatorLine1)
        
        // Check details
        checkDetailsLabel.font = UIFont.systemFont(ofSize: 14)
        checkDetailsLabel.text = NSLocalizedString("Check the details", comment: "")
        checkDetailsLabel.textColor = UIColor.black
        cardView.addSubview(checkDetailsLabel)
        
        // Bottom right image
        if #available(iOS 13.0, *) {
            bottomRightImg.image = UIImage(systemName: "chevron.right")
        } else {
            bottomRightImg.image = UIImage(named: "more")
        }
        cardView.addSubview(bottomRightImg)
    }
    
    override func layoutSubviews() {
        super.layoutSubviews()
        
        let cardWidth = frame.width - 20
        let cardHeight = frame.height - 50
        
        timeLabel.frame = CGRect(x: 0, y: 20, width: frame.width - 20, height: 20)
        cardView.frame = CGRect(x: 10, y: 50, width: cardWidth, height: cardHeight)
        
        appIconView.frame = CGRect(x: 10, y: 10, width: 25, height: 25)
        appNameLabel.frame = CGRect(x: 45, y: 13, width: cardWidth - 55, height: 20)
        topRightImg.frame = CGRect(x: cardWidth - 45, y: 15, width: 35, height: 8)
        
        separatorLine.frame = CGRect(x: 0, y: 45, width: cardWidth, height: 0.5)
        titleLabel.frame = CGRect(x: 10, y: 55, width: cardWidth - 20, height: 20)
        
        let contentHeight = calculateContentHeight()
        contentLabel.frame = CGRect(x: 10, y: 80, width: cardWidth - 20, height: contentHeight)
        
        separatorLine1.frame = CGRect(x: 10, y: 85 + contentHeight, width: cardWidth - 20, height: 0.5)
        checkDetailsLabel.frame = CGRect(x: 10, y: 95 + contentHeight, width: 200, height: 20)
        bottomRightImg.frame = CGRect(x: cardWidth - 40, y: 95 + contentHeight, width: 10, height: 15)
    }
    
    private func calculateContentHeight() -> CGFloat {
        let maxSize = CGSize(width: frame.width - 40, height: CGFloat.greatestFiniteMagnitude)
        let contentSize = contentLabel.text?.boundingRect(with: maxSize,
                                                         options: .usesLineFragmentOrigin,
                                                         attributes: [NSAttributedString.Key.font: UIFont.systemFont(ofSize: 14)],
                                                         context: nil).size
        return contentSize?.height ?? 0
    }
    
    func configure(with model: SubscribeInfoModel) {
        appNameLabel.text = model.mnpName
        titleLabel.text = model.tmplTitle
        contentLabel.text = model.content
        
        let date = Date(timeIntervalSince1970: TimeInterval(model.dataTime))
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
        timeLabel.text = formatter.string(from: date)
        
        setNeedsLayout()
    }
}

class TCMPPSubscribeInfoVC: UIViewController, UITableViewDataSource, UITableViewDelegate {
    
    private var tableView: UITableView!
    private var refreshControl: UIRefreshControl!
    private var modelData: [SubscribeInfoModel] = []
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.title = NSLocalizedString("Service Notice", comment: "")
        if #available(iOS 13.0, *) {
            self.view.backgroundColor = UIColor.systemBackground
        } else {
            self.view.backgroundColor = UIColor.white
        }
        
        setupTableView()
        requestData()
    }
    
    private func setupTableView() {
        tableView = UITableView(frame: .zero, style: .plain)
        tableView.dataSource = self
        tableView.delegate = self
        tableView.separatorStyle = .none
        tableView.backgroundColor = UIColor(red: 234/255.0, green: 234/255.0, blue: 234/255.0, alpha: 1)
        tableView.register(SubscribeInfoCell.self, forCellReuseIdentifier: "SubscribeInfoCellID")
        view.addSubview(tableView)
        
        tableView.translatesAutoresizingMaskIntoConstraints = false
        NSLayoutConstraint.activate([
            tableView.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            tableView.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            tableView.trailingAnchor.constraint(equalTo: view.trailingAnchor),
            tableView.bottomAnchor.constraint(equalTo: view.bottomAnchor)
        ])
        
        refreshControl = UIRefreshControl()
        refreshControl.addTarget(self, action: #selector(refreshData), for: .valueChanged)
        tableView.refreshControl = refreshControl
    }
    
    @objc private func refreshData() {
        requestData()
    }
    
    private func requestData() {
        guard let token = TCMPPUserInfo.shared.token else {
            showToast(NSLocalizedString("Please login first", comment: ""))
            refreshControl.endRefreshing()
            return
        }
        let appId = TMFMiniAppSDKManager.sharedInstance().getConfigAppKey()
        if appId.isEmpty {
            showToast(NSLocalizedString("AppId is empty", comment: ""))
            refreshControl.endRefreshing()
            return
        }
        
        TCMPPDemoLoginManager.shared.getMessage(token: token, appId: appId, offset: 0, success: { [weak self] messages in
            DispatchQueue.main.async {
                self?.refreshControl.endRefreshing()
                
                if messages.isEmpty {
                    self?.showToast(NSLocalizedString("No messages", comment: ""))
                    return
                }
                
                self?.modelData = messages.map { SubscribeInfoModel(jsonData: $0) }
                self?.tableView.reloadData()
            }
        }, failure: { [weak self] error in
            DispatchQueue.main.async {
                self?.refreshControl.endRefreshing()
                self?.showToast(error.localizedDescription)
            }
        })
    }
    
    private func showToast(_ message: String) {
        let icon = UIImage(named: "success")
        let toast = ToastView(icon: icon!, title: message)
        toast.show(withDuration: 2.0)
    }
    
    // MARK: - UITableViewDataSource
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return modelData.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "SubscribeInfoCellID", for: indexPath) as! SubscribeInfoCell
        cell.configure(with: modelData[indexPath.row])
        return cell
    }
    
    // MARK: - UITableViewDelegate
    
    func tableView(_ tableView: UITableView, heightForRowAt indexPath: IndexPath) -> CGFloat {
        let content = modelData[indexPath.row].content
        let maxSize = CGSize(width: view.frame.width - 40, height: CGFloat.greatestFiniteMagnitude)
        let contentSize = content.boundingRect(with: maxSize,
                                             options: .usesLineFragmentOrigin,
                                             attributes: [NSAttributedString.Key.font: UIFont.systemFont(ofSize: 14)],
                                             context: nil).size
        
        return contentSize.height + 180
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        tableView.deselectRow(at: indexPath, animated: true)
        
        let model = modelData[indexPath.row]
        var verType: TMAVersionType = .online
        
        if model.state == "developer" {
            verType = .develop
        } else if model.state == "trial" {
            verType = .preview
        }
        
        TMFMiniAppSDKManager.sharedInstance().startUpMiniApp(withAppID: model.mnpId,
                                                            verType: verType,
                                                            scene: .aioEntry,
                                                            firstPage: model.page,
                                                            paramsStr: nil,
                                                            parentVC: self) { [weak self] error in
            if let error = error {
                self?.showErrorInfo(error)
            }
        }
    }
    
    private func showErrorInfo(_ error: Error) {
        let errorMsg: String
        if !error.localizedDescription.isEmpty {
            errorMsg = "\(error.localizedDescription)\n\(error._code)\n\(error.localizedDescription)"
        } else {
            errorMsg = "\(error.localizedDescription)\n\(error._code)"
        }
        
        let alert = UIAlertController(title: "Error", message: errorMsg, preferredStyle: .alert)
        present(alert, animated: true)
        
        DispatchQueue.main.asyncAfter(deadline: .now() + 2.0) {
            alert.dismiss(animated: true)
        }
    }
} 